### PR TITLE
Support Gather for variable inputs

### PR DESCRIPTION
**Pull request by l-bat**
relates: #17647 
**merge with extra:** opencv/opencv_extra#774

<cut/>

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.2.0:16.04
build_image:Custom Win=openvino-2020.3.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -1342,32 +1342,64 @@ void ONNXImporter::populateNet(Net dstNet)
...
+                    opencv_onnx::NodeProto proto;
+                    proto.add_input(node_proto.input(0));
+                    proto.add_output(sliceLp.name);
+                    addLayer(dstNet, sliceLp, proto, layer_id, outShapes);
+
+                    inpShape.erase(inpShape.begin() + axis);
+                    layerParams.type = "Reshape";
+                    layerParams.set("dim", DictValue::arrayInt(&inpShape[0], inpShape.size()));
+                    node_proto.set_input(0, sliceLp.name);
+                }
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1397]: Can you please describe why we need Reshape?

Comment (l-bat) [reply modules/dnn/src/onnx/onnx_importer.cpp:1397]: I am trying to describe it on example:
```python
 class Gather(nn.Module):
    def forward(self, x):
        return x[..., 1] # gather
```
Gather op is used to index variable input. First, we need to get Slice and then squeeze corresponding axis.

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1397]: That means every Gather over tensor with dims > 1 will produce `dims - 1` dimensional tensor? Please add it in the comments.

Does it also mean that all existing tests for Gather covers only 1D case?

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1397]: In that scenario, which dimensions we covered before without reshaping?

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1397]: Ok, git it, thanks!

**Pull request merged**